pkg/kubelet/runtimeregistry/kuberuntime_registry.go

```diff
@@ -0,0 +1,295 @@
+	return r.RuntimeServices, nil
+}
+
+func (r *KubeRuntimeRegistry) GetAllImageServices() ( map[string]*ImageService, error) {
+	return r.ImageServices, nil
+}
+
+func (r *KubeRuntimeRegistry) GetPrimaryRuntimeService() (*RuntimeService, error) {
+	for _, runtimeService := range r.RuntimeServices {
+		if runtimeService.IsPrimary == true {
+			return runtimeService, nil
+		}
+	}
+
+	return nil, fmt.Errorf("primary runtime servcie is not defined")
+}
+
+func (r *KubeRuntimeRegistry) GetRuntimeServiceByWorkloadType(workloadtype string) (*RuntimeService, error) {
+	for _, runtimeService := range r.RuntimeServices {
+		if runtimeService.WorkloadType == workloadtype {
+			return runtimeService, nil
+		}
+	}
+
+	return nil, fmt.Errorf("runtime servcie for workload type %v is not defined", workloadtype)
+}
+
+func (r *KubeRuntimeRegistry) GetImageServiceByWorkloadType(workloadtype string) (*ImageService, error) {
+	for _, imageService := range r.ImageServices {
+		if imageService.WorkloadType == workloadtype {
+			return imageService, nil
+		}
+	}
+
+	return nil, fmt.Errorf("image servcie for workload type %v is not defined", workloadtype)
+}
+
+// Get status for all runtime services
+func (r *KubeRuntimeRegistry) GetAllRuntimeStatus() (map[string]map[string]bool, error) {
+	statuses := make(map[string]map[string]bool)
+	vmServices := make(map[string]bool)
+	containerServices := make(map[string]bool)
+
+	for runtimeName, runtimeService := range r.RuntimeServices {
+		workloadType := runtimeService.WorkloadType
+		runtimeReady := true
+
+		status, err := runtimeService.ServiceApi.Status()
+		if err != nil || status == nil {
+			runtimeReady = false
+		}
+
+		for _, c := range status.GetConditions() {
+			if c.Status != true {
+				runtimeReady = false
+				break
+			}
+		}
+
+		if workloadType == "vm" {
+			vmServices[runtimeName] = runtimeReady
+		} else {
+			containerServices[runtimeName] = runtimeReady
+		}
+	}
+
+	statuses["vm"] = vmServices
+	statuses["container"] = containerServices
+
+	return statuses, nil
+}
+
+// remoteRuntimeEndpoint format:  runtimeName:workloadType:endpoint;runtimeName:workloadType:endpoint;
+// first one is assumed as the default runtime service
+// for legacy cluster agent without meeting the current format requirement,
+// "default" will be the runtimeName and "container" will be the workloadType
+// returns error if any of the runtime endpoint is malformatted or cannot be retrieved from remote service endpoint
+// TODO: 1. runtime services set from configmap
+//       2. support image_service_endpoint if needed
+//
+func buildRuntimeServicesMapFromAgentCommandArgs(remoteRuntimeEndpoints string) (map[string]*RuntimeService, map[string]*ImageService, error) {
+	if remoteRuntimeEndpoints == "" {
+		return nil, nil, fmt.Errorf("runtimeEndpoints is empty")
+	}
+
+	runtimeServices := make(map[string]*RuntimeService)
+	imageServices := make(map[string]*ImageService)
+
+	defer func() {
```

Comment:
is this debug print?

Reply:
this is log to list registed runtime services. since it only logged once when kubelet starts. level 4 should be appropriate.